Request JavaScript/ECMAscript JMRI scripting users contact us

bobjacobsen opened the pull request:
When JavaScript/ECMAscript scripts are first run, a warning is posted to the log that requests the user get in touch on the jmriusers list.  Hopefully this will help locate a few sample .js scripts for us...

After 5.10 is out, this might be upgraded to a dialog box warning to bring more attention to this.

devel-bobm commented:
If users come forward with JavaScript examples, will it be necessary to add a dialog box later?  I hope not...

But if a dialog box is added in the future, be sure to allow users to disable that dialog box!

danielb987 commented on java/src/jmri/script/ScriptEngineSelector.java:
```diff
@@ -37,6 +37,10 @@ public Engine getSelectedEngine() {
      */
     public void setSelectedEngine(@Nonnull Engine engine) {
         _selectedEngine = engine;
+        if (engine.getLanguageName().equals(ECMA_SCRIPT)) {
+            log.warn("*** Scripting with JavaScript/ECMAscript is being deprecated.***");
+            log.warn("*** Please contact us on the jmriusers group for assistance. ***");
+        }
```
For me as a developer, it's very clear what this is about. But for a user, it might be confusing or seem to be irrelevant. I suggest adding: `and is about to be removed`.

`*** Scripting with JavaScript/ECMAscript is being deprecated and is about to be removed.***`

danielb987 commented:
> After 5.10 is out, this might be upgraded to a dialog box warning to bring more attention to this.

I think it would be good to have the dialog box already in 5.9.x. I'm a developer and I never looks in the log unless I have a very specific reason to do so...

I agree with @devel-bobm that the dialog box needs the ability to disable it.

KenC57 commented:
I might say 'and will be removed if nobody is using this' or something
like that to make it clear they need to respond or it will go away
later.

danielb987 commented:
> I might say 'and will be removed if nobody is using this' or something like that to make it clear they need to respond or it will go away later.

I agree. Thanks.

bobjacobsen commented:
I've added a dialog box and changed the language to 

>  Scripting with JavaScript/ECMAscript is being deprecated and may soon be removed.  If you are using this, please contact us on the jmriusers group for assistance.

Can somebody please point me to an example of a dialog box that can be disabled?  I know I've seen one of those, but I don't remember where.  Thanks in advance.

devel-bobm commented:
> Can somebody please point me to an example of a dialog box that can be disabled?

Sure.  See class `DeleteBeanWorker` in `jmri.jmrit.beantable.BeanTableDataModel`.  Specifically, see :
```
            if (getDisplayDeleteMsg() == 0x02 && message.toString().isEmpty()) {
                doDelete(bean);
            } else {
                JPanel container = new JPanel();
                container.setBorder(BorderFactory.createEmptyBorder(10, 10, 10, 10));
                container.setLayout(new BoxLayout(container, BoxLayout.Y_AXIS));
                if (count > 0)
...
```

The 'if()' checks if the delete warning has been disabled, and simply does the delete.  The 'else' deals with the dialog box and manages the warning disable state.

bobjacobsen commented:
I've updated this so that the dialog box is shown when a JavaScript/ECMAscript script is first executed, then it's silent until the next run of the program.  It does _not_ remember that it's been shown from one run to another.  The idea is to really, really encourage people to get in touch to learn how to turn that off.  

To turn the warning off, there's a new script that can be run at startup.  

Normally, handling options using startup-scripts is inconvenient, but I'm OK with that inconvenience because AFAIK nobody is ever going to encounter it.

danielb987 commented on java/src/jmri/script/ScriptEngineSelector.java:
```diff
@@ -114,6 +114,15 @@ public boolean isJython() {
+         * Is this engine for JavaScript/ECMAscript?
+         * @return true if JavaScript/ECMAscript, false otherwise
+         */
+        public boolean isJavaScript() {
```
No need to change. But what's the purpose of this method? It doesn't seem to be used anywhere.

bobjacobsen commented:
In an earlier version of this PR, I was using that method to identify whether the warning should be triggered or not. This version of the PR works at a lower level to make the detection happen when the actual execution of a script takes place, not when it's first selected. Since I'd already written this method, I just left it in place in case it's useful later on.
